### DISCUSSION TITLE
What is mctLimitState?

### DISCUSSION THREAD

**Discussion by marcelo-earth** [nasa/openmct#5775] [Community Q&A]
Hi OpenMCT community!

I can't find the definition of `mctLimitState` in the documentation, and I see that it is used in the code. I ask this because opening the object that brings back historical telemetry data, I see that there is a new property called `mctLimitState`.

What is it about? Is it okay for it to be `undefined`?

From `openmct-tutorial`:
```js
function HistoricalTelemetryPlugin() {
    return function install (openmct) {
        var provider = {
            supportsRequest: function (domainObject) {
                return domainObject.type === 'example.telemetry';
            },
            request: function (domainObject, options) {
                var url = '/history/' +
                    domainObject.identifier.key +
                    '?start=' + options.start +
                    '&end=' + options.end;

                return http.get(url)
                    .then(function (resp) {
                        console.log(resp.data); // 👈
                        return resp.data;
                    });
            }
        };

        openmct.telemetry.addProvider(provider);
    }
}
```

<img width="593" alt="image" src="https://user-images.githubusercontent.com/47821093/190537621-c1c5141c-0fc3-453a-a82e-828b191f9eb0.png">

**Comment by jvigliotta** [comment]
Hi! What version of Open MCT is the tutorial using when you see this? Also, what branch/step of the tutorial are you on? You can find the Open MCT version by clicking the Open MCT login in the top right. Thanks!

**Reply by marcelo-earth** [reply]
Hi @jvigliotta ! I'm using **Version: 2.0.5**, and the branch is `completed`. I just write a `console.log` in the `HistoricalTelemetryPlugin` as I show in the first post. The result is the same:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/47821093/196834832-d910f8b2-f20e-4b54-81e2-4a37043316c3.png">

**Comment by jvigliotta** [comment]
After further investigation and a little help from @khalidadil (thanks!) I see that this is added here: https://github.com/nasa/openmct/blob/master/src/plugins/plot/configuration/PlotSeries.js#L431 So, it's not coming back from the server, but when viewing the telemetry data in a plot, each datum get's this property. It shouldn't be any issue if it is undefined. It's used by plots to show limit information related to those data.

**Reply by marcelo-earth** [reply]
Thanks @jvigliotta and @khalidadil!
I understand. Yes, it works well even if it's `undefined`.

>  when viewing the telemetry data in a plot, each datum get's this property.
>  It's used by plots to show limit information related to those data.

I have a doubt. **What does `mctLimitState` indicate about each datum?** What does `mctLimitState` contains when it's not `undefined`.

Again, so much thanks 🙏 .

**Comment by jvigliotta** [comment]
@marcelo-earth very sorry for the very long delay, but if you were still wondering and for anyone else interested, if there is a limit provider for the telemetry then a limit state object that contains styling information is returned. From the example in [SinewaveLimitProvider.js](https://github.com/nasa/openmct/blob/master/example/generator/SinewaveLimitProvider.js), we can see the structure:

```js
LIMITS = {
  rh: {
    cssClass: 'is-limit--upr is-limit--red',
    low: RED,
    high: Number.POSITIVE_INFINITY,
    name: 'Red High'
  },
  rl: {
    cssClass: 'is-limit--lwr is-limit--red',
    high: -RED,
    low: Number.NEGATIVE_INFINITY,
    name: 'Red Low'
  },
  yh: {
    cssClass: 'is-limit--upr is-limit--yellow',
    low: YELLOW,
    high: RED,
    name: 'Yellow High'
  },
  yl: {
    cssClass: 'is-limit--lwr is-limit--yellow',
    low: -RED,
    high: -YELLOW,
    name: 'Yellow Low'
  }
};
```

The [limitUtils.js](https://github.com/nasa/openmct/blob/master/src/plugins/plot/chart/limitUtil.js) provides a utility to extract the cssClass:

```js
export function getLimitClass(limit, prefix) {
  let cssClass = '';
  //If color exists then use it, fall back to the cssClass
  if (limit.color) {
    cssClass = `${cssClass} ${prefix}${limit.color}`;
  } else if (limit.cssClass) {
    cssClass = `${cssClass}${limit.cssClass}`;
  }

  // If we applied the cssClass then skip these classes
  if (limit.cssClass === undefined) {
    if (limit.isUpper) {
      cssClass = `${cssClass} ${prefix}upr`;
    } else {
      cssClass = `${cssClass} ${prefix}lwr`;
    }

    if (limit.level) {
      cssClass = `${cssClass} ${prefix}${limit.level}`;
    }

    if (limit.needsHorizontalAdjustment) {
      cssClass = `${cssClass} --align-label-right`;
    }

    if (limit.needsVerticalAdjustment) {
      cssClass = `${cssClass} --align-label-below`;
    }
  }

  return cssClass;
}
```

These classes are defined here: [_limits.scss](https://github.com/nasa/openmct/blob/master/src/styles/_limits.scss)

These styles can then be applied to plot legends, table cells and I believe table rows based on the values of the telemetry.
